### ISSUE TITLE
SET_TYPEOF

### ISSUE THREAD

**Issue by bbolker** [RcppCore/Rcpp#1312]
With the usual apologies if this is a red herring: the `lme4` package is getting 

> Found non-API call to R: ‘SET_TYPEOF’. Compiled code should not call non-API entry points in R.

NOTEs with r-devel. My best guess is that these are coming upstream from Rcpp, e.g. see [here](https://github.com/search?q=repo%3ARcppCore%2FRcpp%20SET_TYPEOF&type=code); I don't find any instances of `SET_TYPEOF` in the `lme4`, `Matrix`, or `RcppEigen` code ...

**Comment by eddelbuettel** [comment]
We appear to have three uses of it:

```sh
edd@ahmad:~/git/rcpp(master)$ ag -c SET_TYPEOF
inst/include/Rcpp/Pairlist.h:1
inst/include/Rcpp/Language.h:1
inst/include/Rcpp/r_cast.h:1
edd@ahmad:~/git/rcpp(master)$ 
```

But I do not think I can offer you anything in the short term. Plead with CRAN that it is not you...

**Comment by Enchufa2** [comment]
Yes, it's clear that we will have to address many of these, but of course it takes time. I don't think CRAN can expect to raise those NOTEs and WARNINGs and have them fixed overnight. I hope they can live with them for a few months in downstream packages to give us here time to adapt, address them one at a time and test properly.

**Comment by Enchufa2** [comment]
Yes, I figured that much. :) Just stating the obvious so that you have a reference to plead with CRAN. ;-)

**Comment by bbolker** [comment]
I don't know if anyone can help me here ... I am *still* seeing SET_TYPEOF notes (on win-builder r-devel [but *not* r-release], and on my local machine that has current Rcpp and RcppEigen installed)

> * checking compiled code ... NOTE
> File 'lme4/libs/x64/lme4.dll':
>  Found non-API call to R: 'SET_TYPEOF'
> Compiled code should not call non-API entry points in R.

This is despite the fact that `SET_TYPEOF` is nowhere in the lme4 source code, and Rcpp and RcppEigen (and Matrix) are the only packages that we are linking to.

Any advice on how to further diagnose/debug this would be welcome ...

**Comment by eddelbuettel** [comment]
What is 'current' Rcpp?  The CRAN version?  That is *expected* as the change taking care of this was made *after* the current CRAN release 1.0.13 was made (and the required out-of-band hotfix 1.0.13-1 did not go there, because that is not what hotfixes do).

If you look e.g. at the laundry list issues "my" packages have at CRAN, you see I get it too in (IIRC) two packages.  

This *will go away* once 1.0.14 hits CRAN, and that is 'any day now' but I am waiting for some dust to settle from the `BH` update in December with three affected packages on the 'fix or else' deadline for next week. 

So in short install Rcpp 1.0.13.6 from the main branch in this repo, or from r-universe, and assert that the warning goes away.  We are sorry about any inconience cause but as you know too, packages with a lot of reverse depends move slowly.

**Comment by bbolker** [comment]
Thanks for the quick response! I looked too quickly at the date of the fix (July 2024) and the date of the current CRAN version (Nov 2024) and didn't realize it was out-of-band.  As long as it will be resolved eventually (and I can still argue with CRAN that it's not my fault), I'm happy.

**Comment by eddelbuettel** [comment]
Yes. I have been updating a few packages that e.g. were due for a removal of an old 'use C++11 for compilation' setting (RcppBDT is one) and have not bothered sending those to CRAN to avoid having to debate this.  If you have r-devel handy (or want to use either my rocker/r-devel or rocker/drd container) it is relatively cheap and easy to that a) yep this arises under 1.0.13 (ie the July 2024 release predating this very PR) but b) not under 1.0.13.6 with is the current tip of the branch having this PR.  So we all bow in the direction of @kevinushey and thank him and once I shepherd this to CRAN in one to two weeks (depending on the speed of the usual email resolutions) we should have _much_ smoother sailing.

**Comment by eddelbuettel** [comment]
> and didn't realize it was out-of-band

This six-month pattern is very much a 'Rcpp-special' but we have been at it for a few years and it a) works and b) predictability is good.
